Kustomization reconciliation failed: too many fields within:

I've installed Flux v0.23.0 and some kustomizations that should be working cannot be updated giving me this error:
2021-11-15T08:40:40.715Z error Kustomization/rights.flux-system - Reconciliation failed after 23.465925ms, next try in 5m0s too many fields within: exploitation_clusterRole_rbac.authorization.k8s.io

As far I can search, the error message is ent by apply mechanism when the object metadata doesn't have the correct format (in this case, the "_" between clusterRole and rbac whould e replaced by ".".

An poor-man bissect give me that the faulty release is the 0.15.0 of kustomize-controller (and I may put another penny on https://github.com/fluxcd/kustomize-controller/commit/3a03d235c241e097e0b6c5d293eca1e55e7c3917 without any validation).

A clean install of 0.23.0 of flux cd, followed by a downgrade of kustomize-controller to version 0.14.1 fixes the issue.

Hi @stefanprodan @matheuscscp I've traced this to its root cause. The fix could be applied in [`fluxcd/cli-utils`](https://github.com/fluxcd/cli-utils).

#### Root Cause

In [`pkg/object/objmetadata.go`](https://github.com/fluxcd/cli-utils/blob/main/pkg/object/objmetadata.go), `ParseObjMetadata()` serializes/deserializes resource inventory entries as `<namespace>_<name>_<group>_<kind>`. The parser correctly extracts the name by working inward from the edges (`strings.Index` for namespace, `strings.LastIndex` for kind, then group). However, after correctly extracting the name, there's an explicit check that **rejects names containing underscores**:

```go
if strings.Contains(name, fieldSeparator) {
    return NilObjMetadata, fmt.Errorf("too many fields within: %s", s)
}
```

This check looks like a false invariant. Kubernetes does allow underscores in resource names. The parsing logic is already correct without this check because API groups use dots and kinds are CamelCase, so underscores in the name wouldn't cause ambiguity.

The bug only manifests on the **second reconciliation**: `ObjMetadata.String()` serializes the name with underscores fine, but when `inventory.Diff()` → `ListMetadata()` → `ParseObjMetadata()` tries to deserialize it on the next reconcile, the check rejects it.

#### Proposed Fix (3 lines removed in `cli-utils`)

```diff
--- a/pkg/object/objmetadata.go
+++ b/pkg/object/objmetadata.go
@@ -91,10 +91,6 @@ func ParseObjMetadata(s string) (ObjMetadata, error) {
 	// Finally, second field name. Name may contain colon transcoded as double underscore.
 	name := s[:index]
 	name = strings.ReplaceAll(name, colonTranscoded, ":")
-	// Check that there are no extra fields by search for fieldSeparator.
-	if strings.Contains(name, fieldSeparator) {
-		return NilObjMetadata, fmt.Errorf("too many fields within: %s", s)
-	}
 	// Create the ObjMetadata object from the four parsed fields.
 	id := ObjMetadata{
 		Namespace: namespace,
```

#### Verification

**1. Unit test with the exact issue payload** (`cli-utils`):

Tested with the exact inventory string from this issue: `_exploitation_clusterRole_rbac.authorization.k8s.io_ClusterRole`

```
=== RUN   TestIssue491_ExactPayload
--- PASS: TestIssue491_ExactPayload (0.00s)
PASS
ok  	github.com/fluxcd/cli-utils/pkg/object	0.009s
```

Confirms parsing succeeds and the round-trip (`String()` → `ParseObjMetadata()`) produces identical results.

**2. Full cli-utils test suite** (all existing tests pass):

```
ok  	github.com/fluxcd/cli-utils/pkg/flowcontrol         1.089s  coverage: 76.5%
ok  	github.com/fluxcd/cli-utils/pkg/kstatus/polling/...  (all pass)
ok  	github.com/fluxcd/cli-utils/pkg/kstatus/status       1.102s  coverage: 82.8%
ok  	github.com/fluxcd/cli-utils/pkg/kstatus/watcher      2.599s  coverage: 51.4%
ok  	github.com/fluxcd/cli-utils/pkg/object                1.696s  coverage: 57.5%
```

**3. Integration test in kustomize-controller** (using `go mod edit -replace` to point at the fix):

Created a test manifest with `ClusterRole` named `exploitation_clusterRole` and tested the full inventory pipeline (`AddChangeSet` → `List` → `ListMetadata` → `Diff`):

```
=== RUN   TestIssue491_UnderscoreInResourceName
=== RUN   .../List_succeeds_with_underscore_in_resource_name
=== RUN   .../ListMetadata_succeeds_with_underscore_in_resource_name
=== RUN   .../Diff_succeeds_with_underscore_in_resource_name
=== RUN   .../round-trip_through_inventory_entry_ID
--- PASS: TestIssue491_UnderscoreInResourceName (0.00s)
PASS
ok  	github.com/fluxcd/kustomize-controller/internal/inventory	0.014s
```

**4. Negative proof** (check back in → tests fail):

With the check restored, the underscore tests fail with exactly the error from this issue:

```
=== RUN   TestParseObjMetadata/Name_can_contain_underscores
    objmetadata_test.go:131: Error parsing inventory when it should have worked:
        too many fields within: test_name_apps
--- FAIL
```

#### Summary

To me, the fix looks like a 3-line deletion in `cli-utils`. The parsing algorithm already correctly handles underscores in names. I didn't find a consumer of that check during my exploration. Happy to open a PR if this approach looks good and there are no concerns about things breaking downstream.

--- a/pkg/object/objmetadata.go
+++ b/pkg/object/objmetadata.go
@@ -91,10 +91,6 @@ func ParseObjMetadata(s string) (ObjMetadata, error) {
 	// Finally, second field name. Name may contain colon transcoded as double underscore.
 	name := s[:index]
 	name = strings.ReplaceAll(name, colonTranscoded, ":")
-	// Check that there are no extra fields by search for fieldSeparator.
-	if strings.Contains(name, fieldSeparator) {
-		return NilObjMetadata, fmt.Errorf("too many fields within: %s", s)
-	}
 	// Create the ObjMetadata object from the four parsed fields.
 	id := ObjMetadata{
 		Namespace: namespace,

--- a/pkg/object/objmetadata_test.go
+++ b/pkg/object/objmetadata_test.go
@@ -97,10 +97,29 @@ func TestParseObjMetadata(t *testing.T) {
 			inventory: &ObjMetadata{},
 			isError:   true,
 		},
-		"Too many fields": {
-			invStr:    "test-namespace_test-name_apps_foo_Deployment",
-			inventory: &ObjMetadata{},
-			isError:   true,
+		"Name can contain underscores": {
+			invStr: "test-namespace_test_name_apps_Deployment",
+			inventory: &ObjMetadata{
+				Namespace: "test-namespace",
+				Name:      "test_name",
+				GroupKind: schema.GroupKind{
+					Group: "apps",
+					Kind:  "Deployment",
+				},
+			},
+			isError: false,
+		},
+		"Name can contain multiple underscores": {
+			invStr: "test-namespace_test_name_with_underscores_apps_Deployment",
+			inventory: &ObjMetadata{
+				Namespace: "test-namespace",
+				Name:      "test_name_with_underscores",
+				GroupKind: schema.GroupKind{
+					Group: "apps",
+					Kind:  "Deployment",
+				},
+			},
+			isError: false,
 		},
 	}
 

